controllers/keystoneapplicationcredential_controller.go

```diff
@@ -0,0 +1,435 @@
+		))
+		return ctrl.Result{RequeueAfter: 10 * time.Second}, nil
+	}
+
+	if !keystoneAPI.IsReady() {
+		instance.Status.Conditions.Set(condition.FalseCondition(
+			keystonev1.KeystoneAPIReadyCondition,
+			condition.RequestedReason,
+			condition.SeverityInfo,
+			keystonev1.KeystoneAPIReadyWaitingMessage,
+		))
+		return ctrl.Result{RequeueAfter: 10 * time.Second}, nil
+	}
+
+	instance.Status.Conditions.MarkTrue(
+		keystonev1.KeystoneAPIReadyCondition,
+		condition.ReadyMessage,
+	)
+
+	// If KeystoneAPI is being deleted, just drop finalizer
+	if !instance.DeletionTimestamp.IsZero() && !keystoneAPI.DeletionTimestamp.IsZero() {
+		if controllerutil.RemoveFinalizer(instance, finalizer) {
+			if err := r.Update(ctx, instance); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+		return ctrl.Result{}, nil
+	}
+
+	if !instance.DeletionTimestamp.IsZero() {
+		return r.reconcileDelete(ctx, instance, helperObj)
+	}
+	return r.reconcileNormal(ctx, instance, helperObj, keystoneAPI)
+}
+
+func (r *ApplicationCredentialReconciler) reconcileNormal(
+	ctx context.Context,
+	instance *keystonev1.ApplicationCredential,
+	helperObj *helper.Helper,
+	keystoneAPI *keystonev1.KeystoneAPI,
+) (ctrl.Result, error) {
+	logger := r.GetLogger(ctx)
+
+	// Decide if we need to create or rotate
+	doRotate, msg := needsRotation(instance)
+	if doRotate {
+		logger.Info(msg)
+
+		// Lookup the user ID
+		userID, ctrlResult, err := r.findUserIDAsAdmin(
+			ctx,
+			instance,
+			helperObj,
+			keystoneAPI,
+			instance.Spec.UserName,
+		)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		if ctrlResult != (ctrl.Result{}) {
+			return ctrlResult, nil
+		}
+		logger.Info("Found user ID", "userName", instance.Spec.UserName, "userID", userID)
+
+		// Build a service‐scoped client
+		userOS, userRes, userErr := keystonev1.GetUserServiceClient(
+			ctx, helperObj, keystoneAPI,
+			instance.Spec.UserName,
+			instance.Spec.Secret,
+			instance.Spec.PasswordSelector,
+		)
+		if userErr != nil {
+			return userRes, userErr
+		}
+		if userRes != (ctrl.Result{}) {
+			return userRes, nil
+		}
+
+		// Create new AC in Keystone
+		newACName := fmt.Sprintf("%s-%s", instance.Name, randomSuffix(5))
+		newID, newSecret, expiresAt, err := r.createACWithName(
+			logger, userOS.GetOSClient(), userID, instance, newACName,
+		)
+		if err != nil {
+			logger.Error(err, "Could not create AC in Keystone")
+			instance.Status.Conditions.Set(condition.FalseCondition(
+				condition.ReadyCondition,
+				condition.ErrorReason,
+				condition.SeverityWarning,
+				fmt.Sprintf("Failed to create AC: %s", err.Error()),
+			))
+			return ctrl.Result{}, err
+		}
+		logger.Info("Created new AC in Keystone", "ACID", newID, "expiresAt", expiresAt)
+
+		// Store it in a Secret
+		secretName := fmt.Sprintf("%s-secret", instance.Name)
+		if err := r.storeACSecret(ctx, helperObj, instance, secretName, newID, newSecret); err != nil {
+			return ctrl.Result{}, err
+		}
```

Comment:
The keystone-operator is somewhat unique in the operators in that there isn't a lot of documentation of what it does.  Just looking at the code , for instance, it wasn't completely clear to me what the name of the ac would be, and the name of the secret etc.  I'm sure I could figure it out - but I shouldn't have to work so hard.

It would be great if there were a small doc (maybe in a docs directory like the other operators) which contains a description of what we'd expect this operator to produce.  It doesn't have to be fancy, but it will help future reviewers and coders who want to enhance this feature.